### ISSUE TITLE
Refactor concurrency model for performance and comprehension

### ISSUE THREAD

**Issue by vcschapp** [gogama/incite#9]
# User Story

1. As an application developer using Incite to support my app, I want the best local app performance possible subject only to CloudWatch Logs service limitations.
2. As an Incite developer I want to understand how the heck it works.

# Background

The current `QueryManager` concurrency design is over-complicated and likely makes inefficient use of available CWL resources.

1. The current design is a bit hard to grok since the outer `loop()` basically tries to do one action per iteration and tries to prioritize starting chunks over polling results.
1. An example of an inefficient resource usage case, if you start a query with, say, 10 startable chunks, and let's say your quota limit is 5 TPS *and* that the Insights web service is capable of sub-200 ms responses. Then it will take about 2 seconds to start all the queries, and during this period polling is throttled, which means if one of the queries finishes by this time (or for previewing has intermediate results) then these results will not be processed.

# Proposal

## Idea

Notionally there are now four queues in the `QueryManager`:

1. The priority queue of unfinished `*stream` values (basically queries with unstarted chunks).
2. The ready queue of startable high-priority chunks (fresh chunks, restarts, and split results). (For some reason this is a `*ring.Ring` in the code but probably could be a straight slice.0
3. The polling ring of running chunks.
4. The cancel queue of chunks that need to be cancelled due to a `Close()` call or another chunk in the same query dying unexpectedly.

Roughly speaking the idea is to have separate goroutines for separate jobs and managing the data structures that make sense for them. Something like:

1. A starter goroutine owns the priority queue, the ready queue, and making `StartQuery` calls. It will block if the size of the polling ring plus the cancel queue is bigger than `Parallel`.
2. A stopper goroutine owns the cancel queue which is probably a simple `chan *chunk` and making `StopQuery` calls. It is usually blocked but springs into action when chunks need to be cancelled, and it has the ability to unblock the starter goroutine anytime the sum of the cancel queue and the polling ring drops below `Parallel`.
3. A poller goroutine owns the polling ring and making `GetQueryResults` calls. Based on the outcome of these calls it may mark a `*stream* into error status (starter goroutine will kill it), rotate the polling ring, remove a chunk from the polling ring, and possibly add one or more chunks to the back (front?) of the ready queue.

## Out of Scope

Building on the above separation of concerns, an even more crazy improvement would be to run each individual web service call in its own sub-goroutine forked off by the main goroutine concern-owning goroutine. For *e.g.*, the starter goroutine could fork off a new goroutine for each `StartQuery` call. This would potentially further improve parallelism by, for example, letting us start chunks up to maximum capacity within the first half second rather than taking more than two seconds for it.

This would be nice for a follow-up issue if we can get the main refactor done.

**Comment by vcschapp** [comment]
Refactor commit [6ca5aaa](https://github.com/gogama/incite/commit/6ca5aaa80ae8100dac069b13f908a7d1c02ff549) gets us most of the way there!

The refactor resulted in a **63%** performance improvement in the unit tests on my laptop (they now take **1.7 seconds** for a full run versus 4.7 seconds previously.

**Comment by vcschapp** [comment]
This is done as of commit 836bde2 and will be released in v1.3.0.

I've run about 70K iterations of the tests on Windows (10) and 50K iterations on Linux (AL2) so pretty confident all the easy-to-hit race conditions have been met.
